### PR TITLE
MM-20424: Migrate 'components/channel_notifications_modal/channel_notifications_modal.jsx' and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by sinansonmez**
#### Summary
This PR migrates 'components/channel_notifications_modal/channel_notifications_modal.jsx' and associated tests to TypeScript

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/16150
JIRA: https://mattermost.atlassian.net/browse/MM-20424

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by devinbinnie — APPROVED** [review]
LGTM! Thanks @sinansonmez!

File: components/channel_notifications_modal/channel_notifications_modal.tsx [hunk 1]
```diff
@@ -154,19 +145,14 @@ export default class ChannelNotificationsModal extends React.PureComponent {
...
 
-    handleUpdateDesktopNotifyLevel = (desktopNotifyLevel) => {
-        this.setState({desktopNotifyLevel});
-    }
-
-    handleUpdateDesktopThreadsNotifyLevel = (desktopThreadsNotifyLevel) => {
-        this.setState({desktopThreadsNotifyLevel});
-    }
+    handleUpdateDesktopNotifyLevel = (desktopNotifyLevel: ChannelNotifyProps['desktop']) => this.setState({desktopNotifyLevel});
+    handleUpdateDesktopThreadsNotifyLevel = (desktopThreadsNotifyLevel: ChannelNotifyProps['desktop_threads']) => this.setState({desktopThreadsNotifyLevel});
```
Comment (M-ZubairAhmed) [inline components/channel_notifications_modal/channel_notifications_modal.tsx:149]: nit : empty line between func

File: components/channel_notifications_modal/channel_notifications_modal.tsx [hunk 2]
```diff
@@ -195,23 +179,15 @@ export default class ChannelNotificationsModal extends React.PureComponent {
...
-    handleUpdatePushThreadsNotificationLevel = (pushThreadsNotifyLevel) => {
-        this.setState({pushThreadsNotifyLevel});
-    }
-
-    handleUpdateIgnoreChannelMentions = (ignoreChannelMentions) => {
-        this.setState({ignoreChannelMentions});
-    }
+    handleUpdatePushNotificationLevel = (pushNotifyLevel: ChannelNotifyProps['push']) => this.setState({pushNotifyLevel});
+    handleUpdatePushThreadsNotificationLevel = (pushThreadsNotifyLevel: ChannelNotifyProps['push_threads']) => this.setState({pushThreadsNotifyLevel});
+    handleUpdateIgnoreChannelMentions = (ignoreChannelMentions: ChannelNotifyProps['ignore_channel_mentions']) => this.setState({ignoreChannelMentions});
```
Comment (M-ZubairAhmed) [inline components/channel_notifications_modal/channel_notifications_modal.tsx:184]: nit : empty lines between

File: components/channel_notifications_modal/channel_notifications_modal.tsx [hunk 3]
```diff
@@ -14,47 +13,43 @@ import {IgnoreChannelMentions, NotificationLevels, NotificationSections} from 'u
...
-             */
-            updateChannelNotifyProps: PropTypes.func.isRequired,
-        }),
-    };
-
-    constructor(props) {
+import {Channel, ChannelNotifyProps} from '@mattermost/types/channels';
+import {UserNotifyProps, UserProfile} from '@mattermost/types/users';
+
+import {PropsFromRedux} from '.';
```
Comment (M-ZubairAhmed) [inline components/channel_notifications_modal/channel_notifications_modal.tsx:19]:
```suggestion
import type {PropsFromRedux} from './index';
```
So we dont have cyclic dependency in js

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: packages/types/src/channels.ts [hunk 4]
```diff
@@ -25,6 +25,8 @@ export type ChannelNotifyProps = {
     mark_unread: 'all' | 'mention';
     push: 'default' | 'all' | 'mention' | 'none';
     ignore_channel_mentions: 'default' | 'off' | 'on';
+    desktop_threads?: 'default' | 'all' | 'mention' | 'none';
+    push_threads?: 'default' | 'all' | 'mention' | 'none';
```
Comment (M-ZubairAhmed) [inline packages/types/src/channels.ts:29]: these belong in UserNotifyProps and not in here. cc @koox00

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

File: packages/types/src/channels.ts (hunk 4 above)
Comment (sinansonmez) [reply packages/types/src/channels.ts]: Updated as suggested

**Review by M-ZubairAhmed — APPROVED** [review]
💯 It looks good to me. Thank you 🎉

**Review by jgilliam17 — APPROVED** [review]
Thank you @sinansonmez 
Tested, looks good to merge.
Verified channel notifications are working as before.
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/247448-0ea40e1-onprem-ent-webapp-pr-11150-mm-ee-test:0ea40e1/mochawesome.html)- no PR related failures.

**Pull request merged**